salt/modules/artifactory.py

```diff
@@ -452,6 +452,7 @@ def __save_artifact(artifact_url, target_file, headers):
         checksum_url = artifact_url + ".sha1"
 
         checksum_success, artifact_sum, checksum_comment = __download(checksum_url, headers)
+        artifact_sum = salt.utils.stringutils.to_str(artifact_sum)
```

Comment:
I'm not convinced that this is the correct fix for this problem.

The issue appears to be that `__download` is producing (sometimes) binary data as the result of a `url.read()` command. I'm pretty sure that we should be checking the `Content-type` header for `text/*`, which we can do with 
```
info = url.info()
# python3
main_type = info.get_content_maintype()
# python2
main_type = info.maintype

if main_type == 'text':
    content = salt.utils.stringutils.to_str(content)
```

Granted, this is the only place that's using this `__download`, but that should correctly fix it for all possible uses.

Note that this `__download` code is *also* found in `salt/modules/nexus.py` - so it's possible that it's got the same sort of failure. Not sure if we should extract `__download`. Looks like the original code was from back in 2014 so maybe we should just refactor these usages to use `salt.utils.http.query`?

Reply:
https://stackoverflow.com/a/36882727/344286 for more info about python2/3